auth/DERMethods.m

```diff
@@ -0,0 +1,106 @@
+classdef DERMethods
+    % Class DERMethods contains all methods pertaining to
+    % Distinguished Encoding Rules (DER) and converting them to naked
+    % formats without the DER headers.
+
+    properties (Constant, Access = protected)
+        DER_ECDSA256_PK_HEADER = uint8([48; 89; 48; 19; 6; 7; 42; 134; 72; 206; 61; 2; 1; 6; 8; 42; 134; 72; 206; ...
+                                        61; 3; 1; 7; 3; 66; 0; 4])
+    end
+
+    methods (Static, Access = public)
+
+        function public_key = DER2PK(der_key, key_type)
+            % strips DER key of all headers and outputs naked public
+            % key  [r; s]
+            switch key_type
+                case "ECDSA256"
+                    if ~all(der_key(1:27) == DERMethods.DER_ECDSA256_PK_HEADER)
+                        error('DERMethods:DER2PK:BadHeader', 'Input header was not expected.');
+                    end
+                    public_key = der_key(28:end);
+                otherwise
+                    error('DERMethods:DER2PK:BadKeyType', 'Key type %s not implemented.', key_type);
+            end
+        end
+
+        function der_key = PK2DER(public_key, key_type)
+            % assembles DER signature from naked public key
+            switch key_type
+                case "ECDSA256"
+                    if length(public_key) ~= 64
+                        error('DERMethods:PK2DER:BadPublicKey', 'Input public key invalid.');
+                    end
+                    der_key = [DERMethods.DER_ECDSA256_PK_HEADER; public_key];
+                otherwise
+                    error('DERMethods:PK2DER:BadKeyType', 'Key type %s not implemented.', key_type);
+            end
+        end
+
+        function public_signature = DER2SIG(der_signature, key_type)
+            % strips DER signature of all headers and outputs naked public
+            % key signature [r; s]
+
+            switch key_type
+                case "ECDSA256"
+                    if der_signature(1) ~= uint8(48)
+                        error('DERMethods:DER2SIG:BadSignature', 'Input signature not valid.');
+                    end
+
+                    der_signature = der_signature(3:end);
+
+                    public_signature = [];
+                    for i = 1:2
```

Comment:
this is more me trying to understand this but why do you need to loop through this

Reply:
In the DER format, it generally follows, [number type][number length in bits][number]. 

In this case
[EC Point x][x length][x][EC Point y][y length][y].

Since the length is included in the spec here, the DER itself is not always the same size and therefore the numbers are not ascribed to the same index each time. Which is why I do this loop. I could have constructed an if/else logic to determine x and y number indices, but is scales badly. I would need logic for 256 and 255 bit cases for each number

Reply:
It is definitely a good idea to throw an error if code receives a bad input because the error that results from a dependency would not make any sense to a user...since it is another level deeper. This is why I took the time to write a bunch of error statements on the DER files...cause the resulting java error makes not sense in context.